net/xia/ppal_ether/main.c

```diff
@@ -162,6 +162,296 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr ||
+	    cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len)
+		return -EINVAL;
+
+	id = cfg->xfc_dst->xid_id;
+	dev = cfg->xfc_odev;
+
+	fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+	if (!fxid) {
+		rc = -ENOENT;
+		goto unlock_bucket;
+	}
+	if (fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+	mether = fxid_mether(fxid);
+
+	if (memcmp(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len) &&
+	    !(mether->host_interface == dev)) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+
+	detach_neigh_to_interface(mether);
+	ether_rt_iops->fxid_rm_locked(&bucket, xtbl, fxid);
+	fxid_free(xtbl, fxid);
+	rc = 0;
+
+unlock_bucket:
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	return rc;
+}
+
+static int main_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+			   struct xip_ppal_ctx *ctx, struct sk_buff *skb,
+			   struct netlink_callback *cb)
+{
+	struct nlmsghdr *nlh;
+	u32 portid = NETLINK_CB(cb->skb).portid;
+	u32 seq = cb->nlh->nlmsg_seq;
+	struct rtmsg *rtm;
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+	struct xia_xid dst;
+	struct nlattr *ha_attr;
+	struct rtnl_xia_ether_addrs *rtha;
+
+	nlh = nlmsg_put(skb, portid, seq, RTM_NEWROUTE, sizeof(*rtm),
+			NLM_F_MULTI);
+	if (!nlh)
+		return -EMSGSIZE;
+
+	rtm = nlmsg_data(nlh);
+	rtm->rtm_family = AF_XIA;
+	rtm->rtm_dst_len = sizeof(struct xia_xid);
+	rtm->rtm_src_len = 0;
+	rtm->rtm_tos = 0;
+	rtm->rtm_table = XRTABLE_MAIN_INDEX;
+	rtm->rtm_protocol = RTPROT_UNSPEC;
+	rtm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rtm->rtm_type = RTN_UNICAST;
+	rtm->rtm_flags = 0;
+
+	dst.xid_type = xtbl->fxt_ppal_type;
+	memmove(dst.xid_id, fxid->fx_xid, XIA_XID_MAX);
+	if (unlikely(nla_put(skb, RTA_DST, sizeof(dst), &dst)))
+		goto nla_put_failure;
+
+	/* Hardware addresses. */
+	ha_attr = nla_nest_start(skb, RTA_MULTIPATH);
+	if (!ha_attr)
+		goto nla_put_failure;
+
+	rtha = nla_reserve_nohdr(skb, sizeof(*rtha));
+	if (!rtha)
+		goto nla_put_failure;
+
+	rcu_read_lock();
+
+	rtha->interface_addr_len = mether->host_interface->addr_len;
+	memmove(rtha->interface_addr, mether->ha, rtha->interface_addr_len);
+	rtha->interface_index = mether->host_interface->ifindex;
+
+	rcu_read_unlock();
+
+	/* No attributes. */
+
+	/* Length of rtnetlink header + attributes. */
+	rtha->attr_len = nlmsg_get_pos(skb) - (void *)rtha;
+
+	nla_nest_end(skb, ha_attr);
+
+	nlmsg_end(skb, nlh);
+	return 0;
+
+nla_put_failure:
+	nlmsg_cancel(skb, nlh);
+	return -EMSGSIZE;
+}
+
+/* Call using fxid_free only. */
+void main_free_ether(struct fib_xid_table *xtbl, struct fib_xid *fxid)
+{
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+
+	xdst_free_anchor(&mether->xem_anchor);
+	kfree(mether);
+}
+
+/* ETHER_FIB all table external operations. */
+static const xia_ppal_all_rt_eops_t ether_all_rt_eops = {
+	[XRTABLE_LOCAL_INDEX] = {
+		.newroute = local_newroute,
+		.delroute = list_fib_delroute,
```

Comment:
Could you replace `list_fib_delroute,` with `ether_rt_iops->fib_delroute`? It's a big more general.

Reply:
Actually it gives an error that the initializer element is not constant, probably because the xia_ppal_rt_eops doesn't contain a const delroute whereas the ether_rt_iops is const. So i'm keeping the original list_fib_delroute for now.

Reply:
Fine. Thank you for testing.